### PR TITLE
Fix pulsing the power wire shocking you repeatedly unless you close tgui

### LINKED ISSUE #52538: Pulsing the power wire shocks you repeatedly unless you close tgui
## Reproduction:
Open a door with screwdriver, find and pulse the power wire.

Watch as you get repeatedly shocked in a loop.

Close the tgui window, watch as the shocks stop

workarounds: Leave in game as a punishment for anyone who likes javascript?

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Fixes #52538.

Making functions that look pure unpure isn't a great idea, it turns out.

## Why It's Good For The Game
Going into crit from pulsing the shock wire once isn't very fun.

## Changelog
:cl:
fix: Pulsing the power wire no longer shocks you repeatedly until you close tgui.
/:cl:

### CODE REVIEW

File: code/datums/wires/airlock.dm [hunk 1]
```diff
@@ -16,9 +16,16 @@
...
+	var/obj/machinery/door/airlock/A = holder
+	if (!issilicon(user) && A.isElectrified() && A.shock(user, 100))
+		return
+
+	return ..()
+
 /datum/wires/airlock/interactable(mob/user)
 	var/obj/machinery/door/airlock/A = holder
-	if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))
+	if(!issilicon(user) && A.isElectrified())
```
Comment (AnturK) [inline code/datums/wires/airlock.dm:28]: Without the shock and clause here, people with insulateds will be unable to operate electrified ones.
From quick glance that shock also has power checks which might be necessary here now.

Comment (Jared-Fogle) [reply code/datums/wires/airlock.dm:28]: Good catch. I can move out all the checks from `shock` into its own proc and use that for the wires.

Comment (Jared-Fogle) [reply code/datums/wires/airlock.dm:28]: This has been fixed.

**Comment by trollbreeder** [comment]
Intended. This makes hacking more dangerous as we intend to move away from greytiding

File: code/modules/mob/living/carbon/carbon.dm [hunk 2]
```diff
@@ -1190,3 +1190,24 @@
...
+
+	return TRUE
+
+/// Returns if the carbon is wearing shock proof gloves
+/mob/living/carbon/proc/wearing_shock_proof_gloves()
+	if (gloves)
+		var/obj/item/clothing/gloves/G = gloves
+		return G.siemens_coefficient == 0
+
+	return FALSE
```
Comment (Timberpoes) [inline code/modules/mob/living/carbon/carbon.dm:1213]: Do you take into account things such as Genetics Insulated and Nerve Grounding surgery as part of any of these checks?

Comment (Timberpoes) [inline code/modules/mob/living/carbon/carbon.dm:1213]: Do you take into account things such as Genetics Insulated mutation and Nerve Grounding surgery as part of any of these checks?

Comment (Jared-Fogle) [reply code/modules/mob/living/carbon/carbon.dm:1213]: These are more or less copied and pasted from the old checks, so they were never able to. I can potentially add these in though.

File: code/datums/wires/airlock.dm [hunk 3]
```diff
@@ -16,10 +16,20 @@
...
+
+	return ..()
+
 /datum/wires/airlock/interactable(mob/user)
 	var/obj/machinery/door/airlock/A = holder
-	if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))
-		return FALSE
+	if(!issilicon(user) && A.isElectrified())
+		var/mob/living/carbon/C = user
+		if (!istype(C) || C.should_electrocute(src))
```
Comment (ShizCalev) [inline code/datums/wires/airlock.dm:30]: shouldn't this be !C.should_electrocute()?

File: code/modules/mob/living/carbon/carbon.dm (hunk 2 above)
Comment (Timberpoes) [reply code/modules/mob/living/carbon/carbon.dm:1213]: Yes, please!

Comment (optimumtact) [reply code/modules/mob/living/carbon/carbon.dm:1213]: Don't scope creep features into a fix pr, rule number 1

Comment (Timberpoes) [reply code/modules/mob/living/carbon/carbon.dm:1213]: Read the code before you comment, rule number 0.

/datum/wires/airlock/interactable(mob/user) returns FALSE early if the airlock is electrified and the user's should_electrocute proc returns TRUE. This part is fine.

should_electrocute() only checks for insuls when it comes to sources of shock immunity. It doesn't check for any other type of shock immunity and will return TRUE despite a player being shock immune.

As a result, people who are immune through other sources cannot interact with an electrified door when this proc feeds its return value back into interactable(...).

This is a bug and should ideally be fixed before the PR is merged, otherwise we're gonna have to wait another 24+ hours after it's merged for some poor schmuck like me to fix it after the fact.

This bug has been introduced because the old check for interactibility was `if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))` - A.shock would return FALSE when the person had some form of TRAIT shock immunity (such as from surgery or genetics)

It goes through a chain of calls where a person with one of the immunities granted by traits would take 0 damage, 0 power would be drained from some power cell and an electrocute_mob proc would return 0 for the total energy drained from the cell. This FALSEy value fed through A.shock() causing it to return FALSE and thus the `if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))` check would resolve to FALSE and interactable() would not return early.

In essence, A.shock() was indirectly checking for the appropriate immunities. This functionality is missing in the above code as should_electrocute doesn't do the relevant checks.

tl;dr - A.shock() eventually led to /mob/living/proc/electrocute_act() which would cause a chain of FALSEy returns which eventually led to A.shock() returning FALSE when a player had the appropriate immunities.

should_electrocute() doesn't replicate these checks. As a result, people who are shock immune are prevented from interacting with shocked doors, which is a bug compared to live code where people who are shock immune are able to interact with shocked doors.

oranges should probably read the code more before commenting.

Comment (optimumtact) [reply code/modules/mob/living/carbon/carbon.dm:1213]: There's no need to be rude, your original message made no such detailed claims.

Had you been clearer in communicating originally, there would be no room for misinterpretation.

Comment (Timberpoes) [reply code/modules/mob/living/carbon/carbon.dm:1213]: Apologies for the hostility and rudeness. I think you caught me at a grumpy time.

File: code/datums/wires/airlock.dm (hunk 3 above)
Comment (Rohesie) [inline code/datums/wires/airlock.dm:30]:
```suggestion
		var/mob/living/carbon/carbon_user = user
		if (!istype(carbon_user) || carbon_user.should_electrocute(src))
```
Less single upper-case letter variables, please.

File: code/datums/wires/airlock.dm [hunk 4]
```diff
@@ -16,10 +16,20 @@
 	add_duds(2)
 	..()
 
+/datum/wires/airlock/interact(mob/user)
+	var/obj/machinery/door/airlock/A = holder
+	if (!issilicon(user) && A.isElectrified() && A.shock(user, 100))
```
Comment (Rohesie) [inline code/datums/wires/airlock.dm:21]:
```suggestion
	var/obj/machinery/door/airlock/holder_airlock = holder
	if (!issilicon(user) && holder_airlock.isElectrified() && holder_airlock.shock(user, 100))
```

File: code/modules/mob/living/carbon/carbon.dm [hunk 5]
```diff
@@ -1190,3 +1190,24 @@
...
+	if(!get_powernet_info_from_source(power_source))
+		return FALSE
+
+	return TRUE
+
+/// Returns if the carbon is wearing shock proof gloves
+/mob/living/carbon/proc/wearing_shock_proof_gloves()
+	if (gloves)
+		var/obj/item/clothing/gloves/G = gloves
+		return G.siemens_coefficient == 0
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/carbon.dm:1211]:
```suggestion
		var/obj/item/clothing/gloves/worn_gloves = gloves
		return worn_gloves.siemens_coefficient == 0
```

File: code/modules/power/power.dm [hunk 6]
```diff
@@ -275,54 +275,64 @@
...
-			var/obj/item/clothing/gloves/G = H.gloves
-			if(G.siemens_coefficient == 0)
-				SEND_SIGNAL(M, COMSIG_LIVING_SHOCK_PREVENTED, power_source, source, siemens_coeff, dist_check)
-				return 0		//to avoid spamming with insulated glvoes on
-
+/// Extracts the powernet and cell of the provided power source
+/proc/get_powernet_info_from_source(power_source)
 	var/area/source_area
-	if(istype(power_source, /area))
+	if (istype(power_source, /area))
```
Comment (Rohesie) [inline code/modules/power/power.dm:281]:
```suggestion
	if (isarea(power_source))
```

File: code/modules/power/power.dm [hunk 7]
```diff
@@ -275,54 +275,64 @@
...
 		cell = apc.cell
 		if (apc.terminal)
 			PN = apc.terminal.powernet
 	else if (!power_source)
-		return 0
+		return
 	else
-		log_admin("ERROR: /proc/electrocute_mob([M], [power_source], [source]): wrong power_source")
-		return 0
+		return
```
Comment (Rohesie) [inline code/modules/power/power.dm:303]: Looks like the `else if (!power_source)` check has become redundant, as the proc will return regardless of it being true or false?

File: code/modules/power/power.dm [hunk 8]
```diff
@@ -275,54 +275,64 @@
...
+
+//Determines how strong could be shock, deals damage to mob, uses power.
+//M is a mob who touched wire/whatever
+//power_source is a source of electricity, can be power cell, area, apc, cable, powernet or null
+//source is an object caused electrocuting (airlock, grille, etc)
+//siemens_coeff - layman's terms, conductivity
+//dist_check - set to only shock mobs within 1 of source (vendors, airlocks, etc.)
+//No animations will be performed by this proc.
+/proc/electrocute_mob(mob/living/carbon/M, power_source, obj/source, siemens_coeff = 1, dist_check = FALSE)
+	if(!istype(M) || ismecha(M.loc))
```
Comment (Rohesie) [inline code/modules/power/power.dm:318]:
```suggestion
	if(!iscarbon(M) || ismecha(M.loc))
```
`M` is a terrible variable name for a carbon.

Comment (Jared-Fogle) [reply code/modules/power/power.dm:318]: Definitely. Most of this is just old code moved to new places.
